### PR TITLE
Add Klee on CI

### PR DESCRIPTION
**Pull request by LinerSu**
Two major changes:

- Integrate harnesses generated for Klee into current ones
- Add Klee on GitHub Actions (either use Klee branch at a push or daily schedule)

### CODE REVIEW

File: seahorn/include/utils.h [hunk 1]
```diff
@@ -10,6 +10,16 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#ifndef __KLEE__ // define macros for KLEE
+#define __KLEE__ 0
```
Comment (agurfinkel) [inline seahorn/include/utils.h:14]: remove. if not defined then do not define as anything else

File: seahorn/jobs/byte_buf_from_array/CMakeLists.txt [hunk 2]
```diff
@@ -1,9 +1,12 @@
 add_executable(byte_buf_from_array
   ${AWS_C_COMMON_ROOT}/source/byte_buf.c
   aws_byte_buf_from_array_harness.c)
+target_compile_definitions(byte_buf_from_array
+  PUBLIC KLEE_MAX_SIZE=${KLEE_MAX_SIZE} # only use for compilation
```
Comment (agurfinkel) [inline seahorn/jobs/byte_buf_from_array/CMakeLists.txt:5]: this should not be necessary for every benchmark. Can set globally like other definitions

**Review by agurfinkel — CHANGES_REQUESTED** [review]
still a few comments to address but getting very close to done

**Comment by LinerSu** [comment]
> still a few comments to address but getting very close to done

The main time difference is the assumption made for the bounded array list. I add only one assumption to avoid multiplication overflow. Here is my implementation:
https://github.com/LinerSu/verify-c-common/blob/klee/seahorn/lib/array_list_helper.c#L39-L45

File: seahorn/jobs/byte_buf_from_array/CMakeLists.txt (hunk 2 above)
Comment (LinerSu) [reply seahorn/jobs/byte_buf_from_array/CMakeLists.txt:5]: Will remove in upcoming commits.

File: seahorn/include/utils.h (hunk 1 above)
Comment (LinerSu) [reply seahorn/include/utils.h:14]: Will remove in upcoming commits.

File: seahorn/CMakeLists.txt [hunk 3]
```diff
@@ -251,9 +251,12 @@ macro(sea_add_klee_bc name)
   set(TARGET_KLEE "${name}.klee")
   add_executable(${TARGET_KLEE} ${ARG_UNPARSED_ARGUMENTS})
   target_compile_definitions(${TARGET_KLEE} PUBLIC 
-    MAX_INITIAL_SIZE=${MAX_INITIAL_SIZE}
+    MAX_BUFFER_SIZE=${MAX_BUFFER_SIZE}
+    MAX_STRING_LEN=${MAX_STRING_LEN}
+    KLEE_MAX_SIZE=${KLEE_MAX_SIZE}
     MAX_ITEM_SIZE=${MAX_ITEM_SIZE}
     MAX_INITIAL_ITEM_ALLOCATION=${MAX_INITIAL_ITEM_ALLOCATION}
+    __KLEE__=1
```
Comment (agurfinkel) [inline seahorn/CMakeLists.txt:259]: dont' really need `=1` here

File: seahorn/CMakeLists.txt [hunk 4]
```diff
@@ -251,9 +251,12 @@ macro(sea_add_klee_bc name)
   set(TARGET_KLEE "${name}.klee")
   add_executable(${TARGET_KLEE} ${ARG_UNPARSED_ARGUMENTS})
   target_compile_definitions(${TARGET_KLEE} PUBLIC 
-    MAX_INITIAL_SIZE=${MAX_INITIAL_SIZE}
+    MAX_BUFFER_SIZE=${MAX_BUFFER_SIZE}
```
Comment (agurfinkel) [inline seahorn/CMakeLists.txt:254]: none of these size definitions are needed. They make it very hard to override the sizes later on.
Instead, code should include `config.h` and then inherit the definitions from the original target.

File: seahorn/include/utils.h [hunk 5]
```diff
@@ -10,6 +10,12 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#ifdef __KLEE__ // define macros for KLEE_ASSUME, the assumption for KLEE only
+#define KLEE_ASSUME(X) __VERIFIER_assume(X)
```
Comment (agurfinkel) [inline seahorn/include/utils.h:14]: why `__VERIFIER_assume` and not `klee_assume`?

File: seahorn/jobs/array_list_clear/CMakeLists.txt [hunk 6]
```diff
@@ -2,11 +2,11 @@ add_executable(array_list_clear
   ${AWS_C_COMMON_ROOT}/source/array_list.c
   aws_array_list_clear_harness.c)
 target_compile_definitions(array_list_clear
-  PUBLIC MAX_ITEM_SIZE=2
-  MAX_INITIAL_ITEM_ALLOCATION=9223372036854775808ULL)
+  PUBLIC MAX_ITEM_SIZE=${MAX_ITEM_SIZE}
```
Comment (agurfinkel) [inline seahorn/jobs/array_list_clear/CMakeLists.txt:5]: I believe these definitions are not needed. They are globally set in `config.h` and in `seahorn/CMakeLists.txt`

File: seahorn/lib/array_list_helper.c [hunk 7]
```diff
@@ -21,30 +22,20 @@ void initialize_array_list(struct aws_array_list *const list) {
...
 }
 
-void initialize_bounded_array_list(struct aws_array_list *const list,
-                                const size_t max_initial_size,
-                                const size_t max_initial_item_allocation,
-                                const size_t max_item_size) {
+void initialize_bounded_array_list(struct aws_array_list *const list) {
   list->current_size = nd_size_t();
-  assume(list->current_size <= max_initial_size);
+  // No need for seahorn of assume(list->current_size <= SIZE_MAX);
```
Comment (agurfinkel) [inline seahorn/lib/array_list_helper.c:27]: remove this comment

File: seahorn/lib/bounds.c [hunk 8]
```diff
@@ -3,7 +3,7 @@
 
 size_t sea_max_buffer_size(void) { return MAX_BUFFER_SIZE; }
 size_t sea_max_string_len(void) { return MAX_STRING_LEN; }
-size_t sea_max_array_list_initial_size(void) { return MAX_INITIAL_SIZE; }
+size_t sea_max_array_list_initial_size(void) { return KLEE_MAX_SIZE; }
```
Comment (agurfinkel) [inline seahorn/lib/bounds.c:6]: either the name of the function is wrong, or it's value. It is called `initial_size`, but returns `KLEE_MAX_SIZE`

**Review by agurfinkel — CHANGES_REQUESTED** [review]
a few minor things that I still have questions about. This looks very good otherwise.

File: seahorn/jobs/array_list_clear/CMakeLists.txt (hunk 6 above)
Comment (LinerSu) [reply seahorn/jobs/array_list_clear/CMakeLists.txt:5]: But the compilation will report an error: "use of undeclared identifier" when I remove it. Should each harness require to include `config.h`? Or is something wrong when those definitions passed to each cmake target?

Comment (agurfinkel) [reply seahorn/jobs/array_list_clear/CMakeLists.txt:5]: this is the place where these definitions should be moved into: https://github.com/seahorn/verify-c-common/blob/master/seahorn/CMakeLists.txt#L309

I can see that these particular ones are missing from there.

Long term -- yes, every harness should include `config.h`, global definition of these macros will be removed. Per instance setting of these parameters will be used to override the default. 

This is what we do for all new unit proofs. So if you are changing these ones, might as well change them to follow the new form. Or, at least, the intermediate step where all definitions are kept in one cmake file

Comment (LinerSu) [reply seahorn/jobs/array_list_clear/CMakeLists.txt:5]: Never mind. Fix it!

**Comment by LinerSu** [comment]
Done the changes based on your comments.

**Review by agurfinkel — APPROVED** [review]

**Comment by agurfinkel** [comment]
waiting for CI, if everything passes, ready to merge

**Pull request merged**